agent/proxycfg/state.go

```diff
@@ -0,0 +1,336 @@
+	if !ok {
+		return nil, errors.New("failed to copy proxy config")
+	}
+
+	return &state{
+		proxyID:  ns.ID,
+		address:  ns.Address,
+		port:     ns.Port,
+		proxyCfg: proxyCfg,
+		token:    token,
+		ch:       make(chan cache.UpdateEvent, 10),
+		snapCh:   make(chan ConfigSnapshot, 1),
+		reqCh:    make(chan chan *ConfigSnapshot, 1),
+	}, nil
+}
+
+// Watch initialised watches on all necessary cache data for the current proxy
+// registration state and returns a chan to observe updates to the
+// ConfigSnapshot that contains all necessary config state. The chan is closed
+// when the state is Closed.
+func (s *state) Watch() (<-chan ConfigSnapshot, error) {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+
+	err := s.initWatches()
+	if err != nil {
+		s.cancel()
+		return nil, err
+	}
+
+	go s.run()
+
+	return s.snapCh, nil
+}
+
+// Close discards the state and stops any long-running watches.
+func (s *state) Close() error {
+	if s.cancel != nil {
+		s.cancel()
+	}
+	return nil
+}
+
+// initWatches sets up the watches needed based on current proxy registration
+// state.
+func (s *state) initWatches() error {
+	// Watch for root changes
+	err := s.cache.Notify(s.ctx, cachetype.ConnectCARootName, &structs.DCSpecificRequest{
+		Datacenter:   s.source.Datacenter,
+		QueryOptions: structs.QueryOptions{Token: s.token},
+	}, rootsWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch the leaf cert
+	err = s.cache.Notify(s.ctx, cachetype.ConnectCALeafName, &cachetype.ConnectCALeafRequest{
+		Datacenter: s.source.Datacenter,
+		Token:      s.token,
+		Service:    s.proxyCfg.DestinationServiceName,
+	}, leafWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch for intention updates
+	err = s.cache.Notify(s.ctx, cachetype.IntentionMatchName, &structs.IntentionQueryRequest{
+		Datacenter:   s.source.Datacenter,
+		QueryOptions: structs.QueryOptions{Token: s.token},
+		Match: &structs.IntentionQueryMatch{
+			Type: structs.IntentionMatchDestination,
+			Entries: []structs.IntentionMatchEntry{
+				{
+					Namespace: structs.IntentionDefaultNamespace,
+					Name:      s.proxyCfg.DestinationServiceName,
+				},
+			},
+		},
+	}, intentionsWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch for updates to service endpoints for all upstreams
+	for _, u := range s.proxyCfg.Upstreams {
+		dc := s.source.Datacenter
+		if u.Datacenter != "" {
+			dc = u.Datacenter
+		}
+
+		switch u.DestinationType {
+		case structs.UpstreamDestTypePreparedQuery:
+			// TODO(banks): prepared queries don't support blocking. We need to come
+			// up with an alternative to Notify that will poll at a sensible rate.
+
+			// err = c.Notify(ctx, cachetype.PreparedQueryName, &structs.PreparedQueryExecuteRequest{
+			//  Datacenter:    dc,
+			//  QueryOptions:  structs.QueryOptions{Token: token},
+			//  QueryIDOrName: u.DestinationName,
+			//  Connect: true,
+			// }, u.Identifier(), ch)
+		case structs.UpstreamDestTypeService:
```

Comment:
If there are only 2 (current) possible values for `DestinationType` it might make more sense to move the `default` body into the `UpstreamDestTypeService` block and hard-fail (return error) if an unexpected value shows up.  That way if a 3rd type is introduced this won't do something unexpected.

Reply:
I wrote it this way because "service" is also the default so it's possible that the field might be empty _implying_ service.

But yeah it would be better to be explicit about accepting empty string or "service" as service and fail on anything completly unexpected.